### ISSUE TITLE
Loss functions

### ISSUE THREAD

**Issue by pboudolf** [hukkelas/DeepPrivacy#18]
Hi,
I am quite new to GANs and experimenting a bit with loss functions. While trying to train my own model I struggled a bit unraveling the code of your losses. Could you please explain what the loss function with id = 1 stands for? 

And is it correct to say that loss function with id 0 represents the wasserstein distance, loss function 2 the true positive rate of the discriminator of real images and loss function 3 the adversarial loss for the generator (fakes scores of discriminator)?

Best regards and thanks in advance!

**Comment by hukkelas** [comment]
Hi, sure!

The loss with Id=1 is the gradient penalty term from the paper "[Improved Training of Wasserstein GANs](https://arxiv.org/abs/1704.00028)". This is to constraint the discriminator properly to be a lipschitz-continuous function. See the paper for further details.

Also, if you're starting out with GANs, I'm not sure if my implementation is the best to refer to, as the optimization process is quite complicated due to training with mixed-precision training using the [APEX framework ](https://github.com/NVIDIA/apex).

**Comment by pboudolf** [comment]
Thank you!
I have one more question. Is there any particular reason why you add exactly 32 channels (that represent the latent space) before starting the upsampling part of the generator? 
More specifically, is there a reason why you don't just add them at the start of the generator or add more or less channels (than 32)?

**Comment by hukkelas** [comment]
There is a paper (can't remember the name of it now sorry) that does experiments with the pix2pix GAN architecture on where to insert the latent variable in encoder-decoder architectures for images.
Their experiments showed that it didn't matter very much where the latent variable was injected; therefore, we decided to insert it at the start of the decoder branch.
The reason for 32x4x4 shape is that it's easy to concatenate with the feature map (which is 4x4).

**Comment by pboudolf** [comment]
That makes sense, thank you very much!

I have one last question that relates to the loss functions again. What exactly is the purpose of this epsilon penalty for the discriminator?
https://github.com/hukkelas/DeepPrivacy/blob/3516f6f3bf2dcb6e7ed4d86195a7c1974077cdf8/deep_privacy/models/loss.py#L76
https://github.com/hukkelas/DeepPrivacy/blob/3516f6f3bf2dcb6e7ed4d86195a7c1974077cdf8/deep_privacy/models/loss.py#L88
Since I suppose the discriminator loss is minimized, shouldn't this term be negative as to make sure the scores of the real images are as high as possible?
Or is it in some way used to restrain the discriminator to learn too fast?

**Comment by hukkelas** [comment]
> Or is it in some way used to restrain the discriminator to learn too fast?

Yeah, it's taken from "Progressive Growing of GANs" paper, simply to restrain the discriminator output size to grow over training.

https://arxiv.org/abs/1710.10196
